### DISCUSSION TITLE
How to group services/service templates together to form an application?

### DISCUSSION THREAD

**Discussion by ghost** [netbox-community/netbox#16831] [Q&A]
So we're looking to try and model the different applications and services that our virutal machines host. For instance, if we have a domain controller, we would want to group all the services like Kerberos, LDAP, DNS, and others into a single "Active Directory" application.

I've been messing around with services and service templates but it seems quite limited. I'm wondering if I'm maybe just doing steps out of order which is what I find I often do when something isn't working how I might expect with Netbox.

One issue I've noticed is that you don't seem to be able to have a service that uses BOTH tcp and udp at the same time. For instance, DNS technically can be either tcp or udp. So when I create a service template for DNS I have to create 2 different templates? one for tcp and one for udp? if that's true, why?

Second, let's say I do indeed create two separate service templates for DNS using both tcp and udp. Now if I want to create a service, it seems to be a 1:1 relationship. I can't select multiple service templates for a service.

Basically, how are you supposed to model an "application" (be it standard things like active directory, or even custom internally developed apps) that use multiple ports and multiple protocols into a single "application" that you can add to a device without having to create 20 entries for each port and each protocol then add those 20 entries individually to a device/virtual machine?

I've asked on reddit in r/netbox with multiple people wondering the same thing with no response. I've also asked in the netbox slack with no response, so this seems like it hasn't been brought up before.

**Comment by jcralbino** [comment]
I am interested user of netbox that would like to better understand how to model applications/ services better

**Comment by ghost** [comment]
I am amazed this isn't getting any traction. I submitted a feature request for this last week and it's seen zero movement as well.
https://github.com/netbox-community/netbox/issues/16841

**Comment by amuckart** [comment]
At the moment I'm treating a 'service' as "IP port/protocol pair that's open on the network". This means I've got `DNS/UDP` and `DNS/TCP`, `LDAP` and `LDAP/SSL` etc. and do have to have lots of them mapped to a given host.

Personally, I like being able to see the "ports open on the network" view, but it would definitely be nice to be able to group them, if for no other reason than it'd make them a lot easier to apply to devices, especially when we've got a standard set that are on everything we build.

**Comment by amuckart** [comment]
With respect to DNS that can be TCP or UDP, I think there's a terminology ambiguity that's worth figuring out too because from a system/user perspective "DNS" is a service but from a network perspective I think these are two different network services.

I'm not sure what the correct solution is, but acknowledging the ambiguity seems like a good start.
